### DISCUSSION TITLE
Installing Netket 3.0 on Apple Silicon Chips

### DISCUSSION THREAD

**Discussion by JoeyT1994** [netket/netket#669] [Q&A] [answered]
I recently installed Netket on a Macbook with an M1 chip - using miniforge3 to install its dependencies and a `pip install --pre netket' for Netket itself.

The version it installed, however, is V2.1.1 and if I uninstall Netket and try to force V3 with pip install netket==3.0b1.post7 I get:

 ERROR: Could not find a version that satisfies the requirement jaxlib>=0.1.57 (from netket) (from versions:none)
 ERROR: No matching distributions found for jaxlib>=0.1.57

Is netket 3.0 dependent on jaxlib but earlier versions of netket are not? I cannot find a version of jaxlib to install in either the conda or pip repositories. Perhaps jaxlib is not supported on the ARM architecture at the moment and this is causing the problem.

**Comment by PhilipVinc** [comment, marked as answer]
> Is netket 3.0 dependent on jaxlib but earlier versions of netket are not? 

yes.

> ERROR: No matching distributions found for jaxlib>=0.1.57

I suspect you installed a native M1 python version.
JaxLib is [not available yet](https://github.com/google/jax/issues/5501) as a _native_ M1 binary, and i think you will have to wait 1 or 2 more months before prebuilt wheels are released.

With that said, I never use conda (and always have problems with it) so i cannot say.
However if you use the x86 version of python you'll be able to `pip install jaxlib` with no problems, as it will be running under rosetta and he thinks it's a standard intel processor. This will allow you to install netket 3. It won't be native but it will run more than fine.

**Comment by JoeyT1994** [comment]
Yes I am currently using a version of python specifically for the M1 architecture.

I will have a go using Rosetta for now then.

Thank you for clearing that up!

**Comment by VolodyaCO** [comment]
I was finally able to install correctly netket using the M1 chip.

I had to:

- Install scipy from source (using numpy 1.21, because numba requires that version, at the time of writing).
- Install jaxlib and jax from source.
- Install llvm locally, version 11 because numba required this.
- Install Python's llvmlite in a particular version that matched numba's requirements.
- Simply run `pip install --pre netket`, and boom!

**Reply by VolodyaCO** [reply]
I am also able to call this m1 version of netket from julia using PyCall. I built PyCall with a particular pipenv python 3.9 environment. Both julia and python are running on m1.

**Reply by PhilipVinc** [reply]
Did you see the instructions on the website? There's an easier way to do this (at least in theory).

All dependencies should work out of the box except numba, which you can install through conda

**Reply by PhilipVinc** [reply]
Anyhow, I'm a month I think numba should be packaged for m1, finally, so the era of all those workaround should finally come to an end.

**Reply by VolodyaCO** [reply]
I don't use conda. I'm a conda foe

**Comment by PhilipVinc** [comment]
For anyone interested, there's now experimental numba wheels for MacOs ARM M1 that can be installed with pip and without conda.

You should run those commands before installing netket. So do something like

```python
pip install numpy importlib-metadata # manually install the NumPy dependeny
pip install -i https://pypi.anaconda.org/numba/label/wheels_experimental_m1/simple numba
pip install netket
```

In the next few weeks those will become officially available and all this mess should finally be solved.

**Reply by PhilipVinc** [reply]
taken from https://github.com/numba/numba/issues/7951

**Comment by erinaldi** [comment]
Following 
```
 pip install 'numpy==1.21.*' importlib-metadata
 pip install -i https://pypi.anaconda.org/numba/label/wheels_experimental_m1/simple numba
 pip install --upgrade pip
 pip install --upgrade 'netket'
```
from the [Get Started widget](https://www.netket.org/get_started/ ) has been tested on MacBook Pro (14-inch, 2021) M1 Pro with macOS Monterey v12.2.

If one installs `open-mpi` with `brew` and adds the MPI dependencies `mpi4jax` and `mpi4py`, the installation can successfully run with multiple processes.

Tested with the `Examples/Ising2d/ising2d.py` code.
